### PR TITLE
HotBackup Tests stubs and sys directory

### LINKED ISSUE #3809: HotBackup stubs
Need hot backup testcase stubs

### PR DESCRIPTION
**Pull request by pgagne**
Closes #3809
-Hotbackup test stubs
-New directory for system tests (tests/foreman/sys) see #3804
-Required .rst files for docs

### CODE REVIEW

File: tests/foreman/sys/test_hot_backup.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+    @stubbed()
+    def test_positive_directory(self):
+        """katello-backup with existing directory
+
+        @id: 3f8285d5-c68b-4d84-8568-bcd7e4f7f19e
+
+        @Steps:
+            1. Populate Satellite system.
+            2. Run katello-backup with existing directory
```
Comment (omaciel) [inline tests/foreman/sys/test_hot_backup.py]: What does it mean 'with existing directory'? Does one have to create a directory first? If so, any specific permissions required? selinux?

File: tests/foreman/sys/test_hot_backup.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+            3. Restore from backup
+
+        @Assert: The backup and restore succeed.
```
Comment (omaciel) [inline tests/foreman/sys/test_hot_backup.py]: Are we talking about asserting that the return code is zero? Any other checks need to be performed, such as checking that the content used to populate the system is still around?

File: tests/foreman/sys/test_hot_backup.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+            4. Restore offline incremenetal backup
+
+        @Assert: Each backup is fully restored.
+
+        """
+        pass
+
+    @stubbed()
+    def test_positive_restore(self):
+        """Restore to a Satellite with config that is been updated since the backup
```
Comment (omaciel) [inline tests/foreman/sys/test_hot_backup.py]: s/is/has

File: tests/foreman/sys/test_hot_backup.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        pass
+
+    @stubbed()
+    def test_positive_load_backup(self):
+        """Load testing, backup
+
+        @id: b0d7de3a-d1d5-4cb9-8041-47301b658408
+
+        @Steps:
+            1. Run a backup while the Satellite is under load
```
Comment (omaciel) [inline tests/foreman/sys/test_hot_backup.py]: Can you define what 'under load' means? Imagine you hand this off to someone to automate, what should the setup do to put the system in this condition?

File: tests/foreman/sys/test_hot_backup.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        pass
+
+    @stubbed()
+    def test_positive_load_restore(self):
+        """Load testing, restore
+
+        @id: c0347aee-48b4-45da-92ef-1e92397c5543
+
+        @Steps:
+            1. Run a restore while the Satellite is under load
```
Comment (omaciel) [inline tests/foreman/sys/test_hot_backup.py]: See previous comment about 'under load'

File: tests/foreman/sys/test_hot_backup.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        pass
+
+    @stubbed()
+    def test_positive_pull_content(self):
+        """Pull content while a backup is running.
+
+        @id: 1026f5f5-4180-4215-acce-876dd2d09ba9
+
+        @Steps:
+            1. Populate a satellite with hosts
```
Comment (omaciel) [inline tests/foreman/sys/test_hot_backup.py]: So hosts but no content? Step `3` seems to indicate that we do need content... custom repo?

File: tests/foreman/sys/test_hot_backup.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    def test_positive_restore(self):
+        """Restore to a Satellite with config that is been updated since the backup
+
+        @id: 03a47dec-9cd4-4cae-88ec-c19dfaa1d5ad
+
+        @Steps:
+            1. Restore online backup
+            2. Restore offline backup
+            3. Restore online incremenetal backup
+            4. Restore offline incremental backu
```
Comment (cswiii) [inline tests/foreman/sys/test_hot_backup.py]: s/backu/backup/

File: tests/foreman/sys/test_hot_backup.py (hunk 2 above)
Comment (pgagne) [inline tests/foreman/sys/test_hot_backup.py]: My first thought was to do a simple check though the cli or ui that all the config and content from the backup got brought in. Do we have any tools that would with this?

File: tests/foreman/sys/test_hot_backup.py (hunk 6 above)
Comment (pgagne) [inline tests/foreman/sys/test_hot_backup.py]: I think this was a typo. What I meant to say was "Populate a satellite" similar to "https://mojo.redhat.com/docs/DOC-1091836". I guess this should be step 1 on each testcase, I am not sure how generic these can be.

File: tests/foreman/sys/test_hot_backup.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+
+class HotBackupTestCase(TestCase):
+    """Implements ``katello-backup --online`` tests"""
+
+    @classmethod
+    def setUpClass(cls):
+        super(HotBackupTestCase, cls).setUpClass()
+
+    @stubbed()
+    def test_positive_directory(self):
```
Comment (sghai) [inline tests/foreman/sys/test_hot_backup.py]: `test_positive_backup_on_existing_directory`

File: tests/foreman/sys/test_hot_backup.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+        """katello-backup with existing directory
+
+        @id: 3f8285d5-c68b-4d84-8568-bcd7e4f7f19e
+
+        @Steps:
+            1. Populate Satellite system.
```
Comment (sghai) [inline tests/foreman/sys/test_hot_backup.py]: empty line is needed between @steps and first point.

File: tests/foreman/sys/test_hot_backup.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+
+    @stubbed()
+    def test_negative_directory(self):
+        """katello-backup with non-existing directory
+
+
+        @id: 946991ad-125a-4f24-a33e-ea27fce38eda
+
+        @Steps:
+            1. Populate Satellite system.
```
Comment (sghai) [inline tests/foreman/sys/test_hot_backup.py]: empty line is needed between @steps and first point and without indent. Like:
```python
@Steps:

 1. Populate Satellite system.
```

File: tests/foreman/sys/test_hot_backup.py [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+        pass
+
+    @stubbed()
+    def test_positive_incremental(self):
+        """Make an incremental backup
+
+        @id: a5b3536f-8365-41c7-9386-cddde588fe8c
+
+        @Steps:
+            1. Configure basic satellite system
```
Comment (sghai) [inline tests/foreman/sys/test_hot_backup.py]: same here empty line and no inident

Comment (sghai) [inline tests/foreman/sys/test_hot_backup.py]: I think step1 should be part of `setup`

File: tests/foreman/sys/test_hot_backup.py [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+
+
+class HotBackupTestCase(TestCase):
+    """Implements ``katello-backup --online`` tests"""
+
+    @classmethod
+    def setUpClass(cls):
+        super(HotBackupTestCase, cls).setUpClass()
+
+    @stubbed()
```
Comment (sghai) [inline tests/foreman/sys/test_hot_backup.py]: @tier3 - I think  all stubs related to backup/restore should fall into tier3 tests. 

- if you'r updating `tier3` decorator then please add another tag in docstrings like ` @CaseLevel: System`
This comment is applicable for all tier3 tests.

**Comment by sghai** [comment]
2 comments for all stubs:
1) stubs should be decorated with @tier3 and following two tags in doc strings
a) @caseautomation: notautomated
b) @CaseLevel: System

2. please run `make docs-clean docs` to generate the docs locally and see if formatting of doc strings is correctly appearing on docs.

File: tests/foreman/sys/test_hot_backup.py (hunk 6 above)
Comment (sghai) [inline tests/foreman/sys/test_hot_backup.py]: As mentioned above, Step1, should be part of `setUpClass`, we will populate satellite once. So `Populate a satellite with hosts/repos/contents` should go under @Setup:

File: tests/foreman/sys/test_hot_backup.py [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+        2. Run ``katello-backup /tmp/backup1``
+        3. Restore from backup
+
+        @Assert: The backup and restore succeed.
+
+        @CaseAutomation: notautomated
+
+        @CaseLevel: System
+        """
+        pass
```
Comment (rochacbruno) [inline tests/foreman/sys/test_hot_backup.py]: `pass` here is not needed at all. python classes accepts only docstring in its body.

```python
class Foo(object):
    """This class does bar
    """
```

is accepted, `pass` is not needed

File: tests/foreman/sys/test_hot_backup.py [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+        1. Ensure the directory /tmp/backup2 does not exist.
+        2. Run ``katello /tmp/backup2``
+
+        @Assert: ``/tmp/backup2`` is created and the backup is saved to it.
+
+        @CaseAutomation: notautomated
+
+        @CaseLevel: System
+        """
+        pass
```
Comment (rochacbruno) [inline tests/foreman/sys/test_hot_backup.py]: the same here

File: tests/foreman/sys/test_hot_backup.py [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+        6. Make config change c2
+        7. run katello-backup <path_to_ib1> --incremental
+
+        @Assert: Backup "ib1" is backed up.
+
+        @CaseAutomation: notautomated
+
+        @CaseLevel: System
+        """
+        pass
```
Comment (rochacbruno) [inline tests/foreman/sys/test_hot_backup.py]: and here

File: tests/foreman/sys/test_hot_backup.py [hunk 16]
```diff
@@ -0,0 +1,228 @@
...
+        3. Restore online incremenetal backup
+        4. Restore offline incremenetal backup
+
+        @Assert: Each backup is fully restored.
+
+        @CaseAutomation: notautomated
+
+        @CaseLevel: System
+        """
+        pass
```
Comment (rochacbruno) [inline tests/foreman/sys/test_hot_backup.py]: and all the others :p

File: tests/foreman/sys/test_hot_backup.py [hunk 17]
```diff
@@ -0,0 +1,228 @@
...
+        @Assert: The restore is successful.
+
+        @CaseAutomation: notautomated
+
+        @CaseLevel: System
+        """
+        pass
+
+    @stubbed()
+    @tier3()
```
Comment (rochacbruno) [inline tests/foreman/sys/test_hot_backup.py]: It is `@tier3` no need to use `()` to call

**Comment by rochacbruno** [comment]
ACK

**Comment by sghai** [comment]
I see multiple commits in this PR. 
@pgagne : Could you please squash them all with a nice commit message and rebase the PR ?

**Comment by pgagne** [comment]
@sghai I can, but do we want to do #3803?

**Comment by sghai** [comment]
@pgagne : Let me try to merge this using `squash & merge`.

**Pull request merged**